### DISCUSSION TITLE
Moeilijkheden linken datastroom

### DISCUSSION THREAD

**Discussion by CelineJansen** [kadaster-labs/sensrnet-home#266] [Q&A] [answered]
Wanneer ik met de API een datastroom wil linken aan een observatie doel (goal) gaat dat prima als het een goal is die aan nog geen enkele andere datastroom gekoppeld is. Als ik een goal wil koppelen aan een datastroom die al een goal heeft, en de betreffende goal is ook al gekoppeld aan een andere datastroom, krijg ik als response een 200 code (ik verwacht dan ook een geslaagde koppeling). In de viewer zie ik dat echter niet terug. 

Overzicht in FME (situatie: 1 device, 2 sensoren, elk met hun eigen datastroom. Datastroom c98f90af-6e59-40a0-8c05-32f04128994f heeft 1 goal, datastroom 3beac715-48b2-4e13-be04-250fb0bb736c heeft er 2, waarvan 1 die ook al aanwezig is voor c98f90af-6e59-40a0-8c05-32f04128994f):
![image](https://user-images.githubusercontent.com/79267472/142264199-f3704b68-f296-470d-b709-0c1e0ba2c96e.png)

Omdat ik in de viewer aan datastroom 3beac715-48b2-4e13-be04-250fb0bb736c maar 1 goal zag hangen. Heb ik de request nog eens gedaan mbv de Swagger UI. Dit is mijn request:

![image](https://user-images.githubusercontent.com/79267472/142263851-890fa4bb-86a1-406c-8188-5c956c0a44ae.png)

De response is nu:

{
  "message": "Object a96fd17a-9795-4021-96ca-5fb5429e7d99 already exists."
}

Maar in de viewer zie ik dat nog steeds niet terug. a96fd17a-9795-4021-96ca-5fb5429e7d99 komt in de viewer alleen maar voor bij c98f90af-6e59-40a0-8c05-32f04128994f en niet bij 3beac715-48b2-4e13-be04-250fb0bb736c.

Dit is de datastroom in de viewer (na verversen van het scherm):
![image](https://user-images.githubusercontent.com/79267472/142265296-9febd964-029c-4f60-ae57-ed12ce1e8a11.png)
Afgaande op bovenstaande response, zou ik bij datastroom a vanaf nu ook observation goal 3 verwachten (naast observation goal 1, die dus wel lukt).

Een GET request op de device geeft het volgende resultaat (wederom zonder a96fd17a-9795-4021-96ca-5fb5429e7d99 bij datastream 3beac715-48b2-4e13-be04-250fb0bb736c (?)):

_{
  "_id": "291a9dc0-b3da-45a4-9bef-be1fb442080d",
  "name": "test device 2",
  "description": "test 1:n, 1 device, 2 sensoren, waarvan 1 met 2 datastromen en 2 observatiedoelen",
  "category": "Sensor",
  "connectivity": "WiFi",
  "datastreams": [
    {
      "theme": [
        "Sound"
      ],
      "observationGoalIds": [
        "a96fd17a-9795-4021-96ca-5fb5429e7d99"
      ],
      "_id": "c98f90af-6e59-40a0-8c05-32f04128994f",
      "sensorId": "0ccd7880-4b2d-4ed2-b550-9f5b02088149",
      "name": "Test datastroom a",
      "description": "DB",
      "unitOfMeasurement": {
        "value": "DB"
      },
      "observationArea": {
        "value": "2m"
      },
      "dataQuality": "(semi-)profesioneel/hoog",
      "isActive": true,
      "isPublic": true,
      "isOpenData": true,
      "containsPersonalInfoData": false,
      "isReusable": true,
      "documentation": "data.eindhoven.nl/home",
      "dataLink": "data.eindhoven.nl/home"
    },
    {
      "theme": [
        "Sound"
      ],
      "observationGoalIds": [
        "e1af7b57-29aa-4e4b-88c5-f270a3e749f8"
      ],
      "_id": "3beac715-48b2-4e13-be04-250fb0bb736c",
      "sensorId": "e5900e9a-161c-447f-a8da-b1f97e47b97b",
      "name": "Test datastroom b",
      "description": "DB",
      "unitOfMeasurement": {
        "value": "DB"
      },
      "observationArea": {
        "value": "2m"
      },
      "dataQuality": "(semi-)profesioneel/hoog",
      "isActive": true,
      "isPublic": true,
      "isOpenData": true,
      "containsPersonalInfoData": false,
      "isReusable": true,
      "documentation": "data.eindhoven.nl/home",
      "dataLink": "data.eindhoven.nl/home"
    }
  ]
}_

Kunnen jullie mij hiermee helpen?

**Comment by marcvanandel** [comment]
Dit lijkt mij 'gewoon' een bug 🤔 @kad-busses , heb jij mogelijkheid (= tijd 😬 ) om hier 's naar te kijken?

**Reply by CelineJansen** [reply]
Dank weer voor jullie snelle reactie, heeft geen grote haast wat mij betreft. Is maar de vraag hoe vaak deze situatie zich in werkelijkheid voor zal doen.

**Reply by kad-busses** [reply]
Bedankt voor alle bijgeleverde informatie, dat helpt met het uitzoeken! Ik lukt me het probleem te reproduceren en kan bevestigen dat het inderdaad een bug in de backend is. Ik heb hier een issue in de backend repo aangemaakt (https://github.com/kadaster-labs/sensrnet-registry-backend/issues/138).

Het lukt me trouwens überhaupt niet om meerdere datastreams binnen dezelfde device/sensor dezelfde observation goal te geven. Het lijkt me echter juist aannemelijk dat meerdere datastreams hetzelfde doel hebben, ze bestaan immers binnen dezelfde device. Hoe vaak de situatie voor zal komen deels af van hoe vaak er meerdere datastreams worden opgevoerd, en dan is het nu wel mogelijk om het doel in elk geval aan 1 datastream te hangen. Lijkt me wel beter om er niet te lang mee te wachten om dit op te lossen.

**Comment by kad-busses** [comment, marked as answer]
De bug is opgelost (https://github.com/kadaster-labs/sensrnet-registry-backend/pull/139) en de nieuwe versie staat live op de demo-omgeving

**Reply by CelineJansen** [reply]
Super, dankjewel. Dan kan ik weer door met testen!
